clerk/verification.go

```diff
@@ -0,0 +1,72 @@
+package clerk
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+const (
+	CookieSession       = "__session"
+	QueryParamSessionId = "_clerk_session_id"
+	OriginHeader        = "Origin"
+)
+
+type VerificationService service
+
+type verifyRequest struct {
+	Token string `json:"token"`
+}
+
+func (s *VerificationService) Verify(req *http.Request) (*Session, error) {
+	if req == nil {
+		return nil, errors.New("cannot verify empty request")
+	}
+	cookie, err := req.Cookie(CookieSession)
+	if err != nil {
+		return nil, errors.New("couldn't find cookie " + CookieSession)
+	}
+
+	sessionToken := cookie.Value
+	sessionId := req.URL.Query().Get(QueryParamSessionId)
+
+	if !isXHR(req) && req.Method == "GET" && sessionId == "" {
+		return s.useClientActiveSession(sessionToken)
+	}
+
+	if sessionId == "" {
+		return nil, errors.New(fmt.Sprintf("no session id is specified via the %s query parameter", QueryParamSessionId))
+	}
+
+	return s.client.Sessions().Verify(sessionId, sessionToken)
+}
+
+func isXHR(req *http.Request) bool {
+	headers := req.Header
+	origin := headers.Get(OriginHeader)
+	return len(origin) > 0
```

Comment:
❓ Is this better than checking for the `X-Requested-With` header? Or maybe we could use both?

Reply:
No idea to be honest! Let's consult with Colin (who came up with this algorithm) later today and modify it accordingly.

Reply:
cc @colinclerk 

Truth is, you can't say with absolute certainty whether a request is an XHR so this is anyway going to be a best-effort approach. I'm just wondering if `X-Requested-With` is the better way of doing this.

Reply:
@agis is right, relying on Origin isn't safe.  Apparently, the Origin is not sent with same-origin requests, and since the SDK handles requests from the developers frontend to the developers backend, it's completely possible the request will be same-origin.

It doesn't look like `X-Requested-With` is safe either, though.  It seems some frameworks may add `X-Requested-With`, but raw fetch and XHR do not.

I put up a proposal for resolution here: https://docs.google.com/document/d/19aUhuOR_zIDti7yWafkGw3wMG026guAObvM_yHe80UM/edit?usp=sharing